src/Microsoft.Unity.Analyzers/GetComponentIncorrectType.cs

```diff
@@ -0,0 +1,82 @@
+/*--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Unity.Analyzers.Resources;
+
+namespace Microsoft.Unity.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class GetComponentIncorrectTypeAnalyzer : DiagnosticAnalyzer
+	{
+		internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			id: "UNT0014",
+			title: Strings.GetComponentIncorrectTypeDiagnosticTitle,
+			messageFormat: Strings.GetComponentIncorrectTypeDiagnosticMessageFormat,
+			category: DiagnosticCategory.TypeSafety,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.GetComponentIncorrectTypeDiagnosticDescription);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		private static readonly HashSet<string> MethodNames = new HashSet<string>(new[]
+		{
+			"GetComponent",
+			"GetComponents",
+			"GetComponentInChildren",
+			"GetComponentsInChildren",
+			"GetComponentInParent",
+			"GetComponentsInParent",
+		});
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzeInvocation, SyntaxKind.InvocationExpression);
+		}
+
+		private static void AnalyzeInvocation(SyntaxNodeAnalysisContext context)
+		{
+			var invocation = (InvocationExpressionSyntax)context.Node;
+			var symbol = context.SemanticModel.GetSymbolInfo(invocation);
+			if (symbol.Symbol == null)
+				return;
+
+			if (!IsGetComponent(symbol.Symbol, out var identifier))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, invocation.GetLocation(), identifier));
+		}
+
+		private static bool IsGetComponent(ISymbol symbol, out string identifier)
```

Comment:
This method has too many responsibilities right now and can be safely split in two separate methods:

1) Checking that the invocation expression is indeed a GetComponent call.
2) Checking that the type argument goes against the rule of the diagnostic.